report_viewer/apps/viewer/middleware/report_views_middleware.py

```diff
@@ -98,22 +94,25 @@ def fingerprint_codename(self, fingerprint_hash: int) -> str:
 
     def __call__(self, request):
         try:
-            absolute_uri: str = request.build_absolute_uri()
-            guid_index = -1
-            if absolute_uri[-1] == "/":
-                guid_index = -2
-
-            guid: str = absolute_uri.split("/")[guid_index]
-            if len(guid) == 36 and UUID(guid, version=4):
-                string_to_hash = self.user_fingerprint(request, SECRET_KEY)
-                fingerprint_hash = self.four_digit_hash(string_to_hash)
-                fingerprint_codename = self.fingerprint_codename(fingerprint_hash)
-                ReportVisitsMetrics(
-                    case=S3Report.objects.get(guid=guid).case,
-                    guid=guid,
-                    fingerprint_hash=fingerprint_hash,
-                    fingerprint_codename=fingerprint_codename,
-                ).save()
+            string_to_hash: str = self.user_fingerprint(request)
+            fingerprint_hash: int = self.four_digit_hash(string_to_hash)
+            if not UserCacheUniqueHash.objects.filter(
+                fingerprint_hash=fingerprint_hash
+            ).exists():
+                absolute_uri: str = request.build_absolute_uri()
+                guid_index = -1
+                if absolute_uri[-1] == "/":
+                    guid_index = -2
+
+                guid: str = absolute_uri.split("/")[guid_index]
```

Comment:
The previous lines are getting the `guid` from the URL. I think that these should be replaced by a utility function with equivalent logic. This function can have a doctring and comprehensive tests.

Also, published report URLs which end in a slash return 404.